{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "36742810",
   "metadata": {},
   "source": [
    "# Labs - Biopython and data formats\n",
    "\n",
    "## Outline\n",
    "\n",
    "- Managing dependencies in Python with environments\n",
    "- Biopython \n",
    "    - Sequences (parsing, representation, manipulation)\n",
    "    - Structures (parsing, representation, manipulation)\n",
    "\n",
    "### 1. Python environments\n",
    "\n",
    "- handles issues with dependencies versions\n",
    "- ensures reproducibility\n",
    "- does not clutter users' global site-packages directory\n",
    "\n",
    "`python3 -m venv venv/       # Creates an environment called venv/`\n",
    "\n",
    "`source venv/bin/activate`\n",
    "\n",
    "`pip install biopython`\n",
    "\n",
    "`pip freeze > requirements.txt`\n",
    "\n",
    "`(venv) % deactivate`\n",
    "\n",
    "On a different machine, the environment can be replicated by creating a new environment and running\n",
    "\n",
    "`pip install -r requirements.txt`\n",
    "\n",
    "### 2. Biopython\n",
    "\n",
    "Biopython is a library consisting of tools for both sequence and structure bioinformatics. Among other things it enables parsing, handling and storing molecular data present in common formats such as FASTA, PDB or mmCIF.\n",
    "\n",
    "Install biopython using `pip install biopython`\n",
    "\n",
    "Functionality divided into packages, list of which is available in the [docs](https://biopython.org/docs/latest/api/Bio.html). \n",
    "\n",
    "Main sequence and structure packages:\n",
    " - [Bio.Seq](https://biopython.org/docs/latest/api/Bio.Seq.html)\n",
    " - [Bio.Align](https://biopython.org/docs/latest/api/Bio.Align.html) \n",
    " - [Bio.SeqIO](https://biopython.org/docs/latest/api/Bio.SeqIO.html)\n",
    " - [Bio.PDB](https://biopython.org/docs/latest/api/Bio.PDB.html) \n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76c2b925",
   "metadata": {},
   "source": [
    " \n",
    "#### Sequences \n",
    " \n",
    " Loading a sequence from a string: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "124d529c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGTACACTG\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "seq = Seq(\"AGTACACTG\")\n",
    "print(seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bb901fb",
   "metadata": {},
   "source": [
    "This creates a [sequence object](https://biopython.org/docs/latest/api/Bio.Seq.html) with a couple of fancy methods, especially when it comes to nuclotide sequences such as `reverse_complement` or `translate`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "61c30fdc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STL\n",
      "AGTACACTG\n",
      "CAGTGTACT\n",
      "CAGUGUACU\n",
      "QCT\n",
      "QCT\n"
     ]
    }
   ],
   "source": [
    "print(seq.translate())\n",
    "print(seq)\n",
    "print(seq.reverse_complement())\n",
    "print(seq.reverse_complement().transcribe())\n",
    "print(seq.reverse_complement().translate())\n",
    "print(seq.reverse_complement().transcribe().translate())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cc4d7807",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAIVMGR*KGAR*\n",
      "MAIVMGR\n",
      "MAIVMGRWKGAR*\n",
      "MAIVMGRWKGAR\n"
     ]
    }
   ],
   "source": [
    "coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
    "print(coding_dna.translate())\n",
    "print(coding_dna.translate(to_stop=True))\n",
    "print(coding_dna.translate(table=2))\n",
    "print(coding_dna.translate(table=2, to_stop=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21a88c4f",
   "metadata": {},
   "source": [
    "Notice, in the example above we used different genetic tables. Check [NCBI genetic codes](https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi) for details.\n",
    "\n",
    "To list all the methods, run, e.g., one of the following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "beb88d44",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['__abstractmethods__', '__add__', '__array_ufunc__', '__bytes__', '__class__', '__contains__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__imul__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__mul__', '__ne__', '__new__', '__radd__', '__reduce__', '__reduce_ex__', '__repr__', '__rmul__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', '_data', 'back_transcribe', 'complement', 'complement_rna', 'count', 'count_overlap', 'defined', 'defined_ranges', 'endswith', 'find', 'index', 'islower', 'isupper', 'join', 'lower', 'lstrip', 'replace', 'reverse_complement', 'reverse_complement_rna', 'rfind', 'rindex', 'rsplit', 'rstrip', 'split', 'startswith', 'strip', 'transcribe', 'translate', 'ungap', 'upper']\n",
      "Help on Seq in module Bio.Seq object:\n",
      "\n",
      "class Seq(_SeqAbstractBaseClass)\n",
      " |  Seq(data, length=None)\n",
      " |  \n",
      " |  Read-only sequence object (essentially a string with biological methods).\n",
      " |  \n",
      " |  Like normal python strings, our basic sequence object is immutable.\n",
      " |  This prevents you from doing my_seq[5] = \"A\" for example, but does allow\n",
      " |  Seq objects to be used as dictionary keys.\n",
      " |  \n",
      " |  The Seq object provides a number of string like methods (such as count,\n",
      " |  find, split and strip).\n",
      " |  \n",
      " |  The Seq object also provides some biological methods, such as complement,\n",
      " |  reverse_complement, transcribe, back_transcribe and translate (which are\n",
      " |  not applicable to protein sequences).\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      Seq\n",
      " |      _SeqAbstractBaseClass\n",
      " |      abc.ABC\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __hash__(self)\n",
      " |      Hash of the sequence as a string for comparison.\n",
      " |      \n",
      " |      See Seq object comparison documentation (method ``__eq__`` in\n",
      " |      particular) as this has changed in Biopython 1.65. Older versions\n",
      " |      would hash on object identity.\n",
      " |  \n",
      " |  __init__(self, data, length=None)\n",
      " |      Create a Seq object.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - data - Sequence, required (string)\n",
      " |       - length - Sequence length, used only if data is None or a dictionary (integer)\n",
      " |      \n",
      " |      You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |      SeqRecord objects, whose sequence will be exposed as a Seq object via\n",
      " |      the seq property.\n",
      " |      \n",
      " |      However, you can also create a Seq object directly:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\")\n",
      " |      >>> my_seq\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
      " |      >>> print(my_seq)\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      \n",
      " |      To create a Seq object with for a sequence of known length but\n",
      " |      unknown sequence contents, use None for the data argument and pass\n",
      " |      the sequence length for the length argument. Trying to access the\n",
      " |      sequence contents of a Seq object created in this way will raise\n",
      " |      an UndefinedSequenceError:\n",
      " |      \n",
      " |      >>> my_undefined_sequence = Seq(None, 20)\n",
      " |      >>> my_undefined_sequence\n",
      " |      Seq(None, length=20)\n",
      " |      >>> len(my_undefined_sequence)\n",
      " |      20\n",
      " |      >>> print(my_undefined_sequence)\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      Bio.Seq.UndefinedSequenceError: Sequence content is undefined\n",
      " |      \n",
      " |      If the sequence contents is known for parts of the sequence only, use\n",
      " |      a dictionary for the data argument to pass the known sequence segments:\n",
      " |      \n",
      " |      >>> my_partially_defined_sequence = Seq({3: \"ACGT\"}, 10)\n",
      " |      >>> my_partially_defined_sequence\n",
      " |      Seq({3: 'ACGT'}, length=10)\n",
      " |      >>> len(my_partially_defined_sequence)\n",
      " |      10\n",
      " |      >>> print(my_partially_defined_sequence)\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      Bio.Seq.UndefinedSequenceError: Sequence content is only partially defined\n",
      " |      >>> my_partially_defined_sequence[3:7]\n",
      " |      Seq('ACGT')\n",
      " |      >>> print(my_partially_defined_sequence[3:7])\n",
      " |      ACGT\n",
      " |  \n",
      " |  ungap(self, gap='-')\n",
      " |      Return a copy of the sequence without the gap character(s) (DEPRECATED).\n",
      " |      \n",
      " |      The gap character now defaults to the minus sign, and can only\n",
      " |      be specified via the method argument. This is no longer possible\n",
      " |      via the sequence's alphabet (as was possible up to Biopython 1.77):\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_dna = Seq(\"-ATA--TGAAAT-TTGAAAA\")\n",
      " |      >>> my_dna\n",
      " |      Seq('-ATA--TGAAAT-TTGAAAA')\n",
      " |      >>> my_dna.ungap(\"-\")\n",
      " |      Seq('ATATGAAATTTGAAAA')\n",
      " |      \n",
      " |      This method is DEPRECATED; please use my_dna.replace(gap, \"\") instead.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __abstractmethods__ = frozenset()\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from _SeqAbstractBaseClass:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Add a sequence or string to this sequence.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> Seq(\"MELKI\") + \"LV\"\n",
      " |      Seq('MELKILV')\n",
      " |      >>> MutableSeq(\"MELKI\") + \"LV\"\n",
      " |      MutableSeq('MELKILV')\n",
      " |  \n",
      " |  __bytes__(self)\n",
      " |  \n",
      " |  __contains__(self, item)\n",
      " |      Return True if item is a subsequence of the sequence, and False otherwise.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_dna = Seq(\"ATATGAAATTTGAAAA\")\n",
      " |      >>> \"AAA\" in my_dna\n",
      " |      True\n",
      " |      >>> Seq(\"AAA\") in my_dna\n",
      " |      True\n",
      " |      >>> MutableSeq(\"AAA\") in my_dna\n",
      " |      True\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Compare the sequence to another sequence or a string.\n",
      " |      \n",
      " |      Sequences are equal to each other if their sequence contents is\n",
      " |      identical:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> seq1 = Seq(\"ACGT\")\n",
      " |      >>> seq2 = Seq(\"ACGT\")\n",
      " |      >>> mutable_seq = MutableSeq(\"ACGT\")\n",
      " |      >>> seq1 == seq2\n",
      " |      True\n",
      " |      >>> seq1 == mutable_seq\n",
      " |      True\n",
      " |      >>> seq1 == \"ACGT\"\n",
      " |      True\n",
      " |      \n",
      " |      Note that the sequence objects themselves are not identical to each\n",
      " |      other:\n",
      " |      \n",
      " |      >>> id(seq1) == id(seq2)\n",
      " |      False\n",
      " |      >>> seq1 is seq2\n",
      " |      False\n",
      " |      \n",
      " |      Sequences can also be compared to strings, ``bytes``, and ``bytearray``\n",
      " |      objects:\n",
      " |      \n",
      " |      >>> seq1 == \"ACGT\"\n",
      " |      True\n",
      " |      >>> seq1 == b\"ACGT\"\n",
      " |      True\n",
      " |      >>> seq1 == bytearray(b\"ACGT\")\n",
      " |      True\n",
      " |  \n",
      " |  __ge__(self, other)\n",
      " |      Implement the greater-than or equal operand.\n",
      " |  \n",
      " |  __getitem__(self, index)\n",
      " |      Return a subsequence as a single letter or as a sequence object.\n",
      " |      \n",
      " |      If the index is an integer, a single letter is returned as a Python\n",
      " |      string:\n",
      " |      \n",
      " |      >>> seq = Seq('ACTCGACGTCG')\n",
      " |      >>> seq[5]\n",
      " |      'A'\n",
      " |      \n",
      " |      Otherwise, a new sequence object of the same class is returned:\n",
      " |      \n",
      " |      >>> seq[5:8]\n",
      " |      Seq('ACG')\n",
      " |      >>> mutable_seq = MutableSeq('ACTCGACGTCG')\n",
      " |      >>> mutable_seq[5:8]\n",
      " |      MutableSeq('ACG')\n",
      " |  \n",
      " |  __gt__(self, other)\n",
      " |      Implement the greater-than operand.\n",
      " |  \n",
      " |  __imul__(self, other)\n",
      " |      Multiply the sequence object by other and assign.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> seq = Seq('ATG')\n",
      " |      >>> seq *= 2\n",
      " |      >>> seq\n",
      " |      Seq('ATGATG')\n",
      " |      \n",
      " |      Note that this is different from in-place multiplication. The ``seq``\n",
      " |      variable is reassigned to the multiplication result, but any variable\n",
      " |      pointing to ``seq`` will remain unchanged:\n",
      " |      \n",
      " |      >>> seq = Seq('ATG')\n",
      " |      >>> seq2 = seq\n",
      " |      >>> id(seq) == id(seq2)\n",
      " |      True\n",
      " |      >>> seq *= 2\n",
      " |      >>> seq\n",
      " |      Seq('ATGATG')\n",
      " |      >>> seq2\n",
      " |      Seq('ATG')\n",
      " |      >>> id(seq) == id(seq2)\n",
      " |      False\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |      Return an iterable of the sequence.\n",
      " |  \n",
      " |  __le__(self, other)\n",
      " |      Implement the less-than or equal operand.\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the length of the sequence.\n",
      " |  \n",
      " |  __lt__(self, other)\n",
      " |      Implement the less-than operand.\n",
      " |  \n",
      " |  __mul__(self, other)\n",
      " |      Multiply sequence by integer.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> Seq('ATG') * 2\n",
      " |      Seq('ATGATG')\n",
      " |      >>> MutableSeq('ATG') * 2\n",
      " |      MutableSeq('ATGATG')\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |      Add a sequence string on the left.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> \"LV\" + Seq(\"MELKI\")\n",
      " |      Seq('LVMELKI')\n",
      " |      >>> \"LV\" + MutableSeq(\"MELKI\")\n",
      " |      MutableSeq('LVMELKI')\n",
      " |      \n",
      " |      Adding two sequence objects is handled via the __add__ method.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return (truncated) representation of the sequence.\n",
      " |  \n",
      " |  __rmul__(self, other)\n",
      " |      Multiply integer by sequence.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> 2 * Seq('ATG')\n",
      " |      Seq('ATGATG')\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return the full sequence as a python string.\n",
      " |  \n",
      " |  back_transcribe(self, inplace=False)\n",
      " |      Return the DNA sequence from an RNA sequence by creating a new Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> messenger_rna = Seq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\")\n",
      " |      >>> messenger_rna\n",
      " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> messenger_rna.back_transcribe()\n",
      " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> sequence = MutableSeq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\")\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence.back_transcribe()\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      >>> sequence.back_transcribe(inplace=True)\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``transcribe`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      Trying to back-transcribe DNA has no effect, If you have a nucleotide\n",
      " |      sequence which might be DNA or RNA (or even a mixture), calling the\n",
      " |      back-transcribe method will ensure any U becomes T.\n",
      " |      \n",
      " |      Trying to back-transcribe a protein sequence will replace any U for\n",
      " |      Selenocysteine with T for Threonine, which is biologically meaningless.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_protein = Seq(\"MAIVMGRU\")\n",
      " |      >>> my_protein.back_transcribe()\n",
      " |      Seq('MAIVMGRT')\n",
      " |  \n",
      " |  complement(self, inplace=None)\n",
      " |      Return the complement as a DNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement()\n",
      " |      Seq('GCT')\n",
      " |      \n",
      " |      Any U in the sequence is treated as a T:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement(inplace=False)\n",
      " |      Seq('GCTAA')\n",
      " |      \n",
      " |      In contrast, ``complement_rna`` returns an RNA sequence:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement_rna()\n",
      " |      Seq('GCUAA')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.complement(inplace=False)\n",
      " |      MutableSeq('GCT')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.complement(inplace=True)\n",
      " |      MutableSeq('GCT')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('GCT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``complement_rna`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  complement_rna(self, inplace=False)\n",
      " |      Return the complement as an RNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement_rna()\n",
      " |      Seq('GCU')\n",
      " |      \n",
      " |      Any T in the sequence is treated as a U:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement_rna()\n",
      " |      Seq('GCUAA')\n",
      " |      \n",
      " |      In contrast, ``complement`` returns a DNA sequence by default:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement()\n",
      " |      Seq('GCT')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.complement_rna()\n",
      " |      MutableSeq('GCU')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.complement_rna(inplace=True)\n",
      " |      MutableSeq('GCU')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('GCU')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``complement_rna`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  count(self, sub, start=None, end=None)\n",
      " |      Return a non-overlapping count, like that of a python string.\n",
      " |      \n",
      " |      The number of occurrences of substring argument sub in the\n",
      " |      (sub)sequence given by [start:end] is returned as an integer.\n",
      " |      Optional arguments start and end are interpreted as in slice\n",
      " |      notation.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"AAAATGA\")\n",
      " |      >>> print(my_seq.count(\"A\"))\n",
      " |      5\n",
      " |      >>> print(my_seq.count(\"ATG\"))\n",
      " |      1\n",
      " |      >>> print(my_seq.count(Seq(\"AT\")))\n",
      " |      1\n",
      " |      >>> print(my_seq.count(\"AT\", 2, -1))\n",
      " |      1\n",
      " |      \n",
      " |      HOWEVER, please note because the ``count`` method of Seq and MutableSeq\n",
      " |      objects, like that of Python strings, do a non-overlapping search, this\n",
      " |      may not give the answer you expect:\n",
      " |      \n",
      " |      >>> \"AAAA\".count(\"AA\")\n",
      " |      2\n",
      " |      >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
      " |      2\n",
      " |      \n",
      " |      For an overlapping search, use the ``count_overlap`` method:\n",
      " |      \n",
      " |      >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
      " |      3\n",
      " |  \n",
      " |  count_overlap(self, sub, start=None, end=None)\n",
      " |      Return an overlapping count.\n",
      " |      \n",
      " |      Returns an integer, the number of occurrences of substring\n",
      " |      argument sub in the (sub)sequence given by [start:end].\n",
      " |      Optional arguments start and end are interpreted as in slice\n",
      " |      notation.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
      " |      3\n",
      " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\"))\n",
      " |      4\n",
      " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\", 3, -1))\n",
      " |      1\n",
      " |      \n",
      " |      For a non-overlapping search, use the ``count`` method:\n",
      " |      \n",
      " |      >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
      " |      2\n",
      " |      \n",
      " |      Where substrings do not overlap, ``count_overlap`` behaves the same as\n",
      " |      the ``count`` method:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"AAAATGA\")\n",
      " |      >>> print(my_seq.count_overlap(\"A\"))\n",
      " |      5\n",
      " |      >>> my_seq.count_overlap(\"A\") == my_seq.count(\"A\")\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(\"ATG\"))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(\"ATG\") == my_seq.count(\"ATG\")\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(Seq(\"AT\")))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(Seq(\"AT\")) == my_seq.count(Seq(\"AT\"))\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(\"AT\", 2, -1))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(\"AT\", 2, -1) == my_seq.count(\"AT\", 2, -1)\n",
      " |      True\n",
      " |      \n",
      " |      HOWEVER, do not use this method for such cases because the\n",
      " |      count() method is much for efficient.\n",
      " |  \n",
      " |  endswith(self, suffix, start=None, end=None)\n",
      " |      Return True if the sequence ends with the given suffix, False otherwise.\n",
      " |      \n",
      " |      Return True if the sequence ends with the specified suffix\n",
      " |      (a string or another Seq object), False otherwise.\n",
      " |      With optional start, test sequence beginning at that position.\n",
      " |      With optional end, stop comparing sequence at that position.\n",
      " |      suffix can also be a tuple of strings to try.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.endswith(\"UUG\")\n",
      " |      True\n",
      " |      >>> my_rna.endswith(\"AUG\")\n",
      " |      False\n",
      " |      >>> my_rna.endswith(\"AUG\", 0, 18)\n",
      " |      True\n",
      " |      >>> my_rna.endswith((\"UCC\", \"UCA\", \"UUG\"))\n",
      " |      True\n",
      " |  \n",
      " |  find(self, sub, start=None, end=None)\n",
      " |      Return the lowest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the lowest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.find(\"AUG\")\n",
      " |      3\n",
      " |      \n",
      " |      The next typical start codon can then be found by starting the search\n",
      " |      at position 4:\n",
      " |      \n",
      " |      >>> my_rna.find(\"AUG\", 4)\n",
      " |      15\n",
      " |  \n",
      " |  index(self, sub, start=None, end=None)\n",
      " |      Return the lowest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the lowest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Raises a ValueError if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.index(\"AUG\")\n",
      " |      3\n",
      " |      \n",
      " |      The next typical start codon can then be found by starting the search\n",
      " |      at position 4:\n",
      " |      \n",
      " |      >>> my_rna.index(\"AUG\", 4)\n",
      " |      15\n",
      " |      \n",
      " |      This method performs the same search as the ``find`` method.  However,\n",
      " |      if the subsequence is not found, ``find`` returns -1 which ``index``\n",
      " |      raises a ValueError:\n",
      " |      \n",
      " |      >>> my_rna.index(\"T\")\n",
      " |      Traceback (most recent call last):\n",
      " |                 ...\n",
      " |      ValueError: ...\n",
      " |      >>> my_rna.find(\"T\")\n",
      " |      -1\n",
      " |  \n",
      " |  islower(self)\n",
      " |      Return True if all ASCII characters in data are lowercase.\n",
      " |      \n",
      " |      If there are no cased characters, the method returns False.\n",
      " |  \n",
      " |  isupper(self)\n",
      " |      Return True if all ASCII characters in data are uppercase.\n",
      " |      \n",
      " |      If there are no cased characters, the method returns False.\n",
      " |  \n",
      " |  join(self, other)\n",
      " |      Return a merge of the sequences in other, spaced by the sequence from self.\n",
      " |      \n",
      " |      Accepts a Seq object, MutableSeq object, or string (and iterates over\n",
      " |      the letters), or an iterable containing Seq, MutableSeq, or string\n",
      " |      objects. These arguments will be concatenated with the calling sequence\n",
      " |      as the spacer:\n",
      " |      \n",
      " |      >>> concatenated = Seq('NNNNN').join([Seq(\"AAA\"), Seq(\"TTT\"), Seq(\"PPP\")])\n",
      " |      >>> concatenated\n",
      " |      Seq('AAANNNNNTTTNNNNNPPP')\n",
      " |      \n",
      " |      Joining the letters of a single sequence:\n",
      " |      \n",
      " |      >>> Seq('NNNNN').join(Seq(\"ACGT\"))\n",
      " |      Seq('ANNNNNCNNNNNGNNNNNT')\n",
      " |      >>> Seq('NNNNN').join(\"ACGT\")\n",
      " |      Seq('ANNNNNCNNNNNGNNNNNT')\n",
      " |  \n",
      " |  lower(self, inplace=False)\n",
      " |      Return the sequence in lower case.\n",
      " |      \n",
      " |      An lower-case copy of the sequence is returned if inplace is False,\n",
      " |      the default value:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_seq = Seq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      Seq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      Seq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      \n",
      " |      >>> my_seq.lower(inplace=True)\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper(inplace=True)\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``lower`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the ``upper`` method.\n",
      " |  \n",
      " |  lstrip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with leading and trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, leading whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.lstrip()\n",
      " |      Seq('ACGT ')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      from the leading end instead.  The order of the characters to be removed\n",
      " |      is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGACGTTACG\").lstrip(\"GCA\")\n",
      " |      Seq('TTACG')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.lstrip(inplace=False)\n",
      " |      MutableSeq('ACGT ')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.lstrip(inplace=True)\n",
      " |      MutableSeq('ACGT ')\n",
      " |      >>> seq\n",
      " |      MutableSeq('ACGT ')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``lstrip`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the strip and rstrip methods.\n",
      " |  \n",
      " |  replace(self, old, new, inplace=False)\n",
      " |      Return a copy with all occurrences of subsequence old replaced by new.\n",
      " |      \n",
      " |      >>> s = Seq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = s.replace(\"AC\", \"XYZ\")\n",
      " |      >>> s\n",
      " |      Seq('ACGTAACCGGTT')\n",
      " |      >>> t\n",
      " |      Seq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      For mutable sequences, passing inplace=True will modify the sequence in place:\n",
      " |      \n",
      " |      >>> m = MutableSeq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = m.replace(\"AC\", \"XYZ\")\n",
      " |      >>> m\n",
      " |      MutableSeq('ACGTAACCGGTT')\n",
      " |      >>> t\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      >>> m = MutableSeq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = m.replace(\"AC\", \"XYZ\", inplace=True)\n",
      " |      >>> m\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      >>> t\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``replace`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  reverse_complement(self, inplace=None)\n",
      " |      Return the reverse complement as a DNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement(inplace=False)\n",
      " |      Seq('TCG')\n",
      " |      \n",
      " |      Any U in the sequence is treated as a T:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").reverse_complement(inplace=False)\n",
      " |      Seq('AATCG')\n",
      " |      \n",
      " |      In contrast, ``reverse_complement_rna`` returns an RNA sequence:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement_rna()\n",
      " |      Seq('UCG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.reverse_complement(inplace=False)\n",
      " |      MutableSeq('TCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.reverse_complement(inplace=True)\n",
      " |      MutableSeq('TCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('TCG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``reverse_complement`` is called on a ``Seq`` object with\n",
      " |      ``inplace=True``.\n",
      " |  \n",
      " |  reverse_complement_rna(self, inplace=False)\n",
      " |      Return the reverse complement as an RNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement_rna()\n",
      " |      Seq('UCG')\n",
      " |      \n",
      " |      Any T in the sequence is treated as a U:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").reverse_complement_rna()\n",
      " |      Seq('AAUCG')\n",
      " |      \n",
      " |      In contrast, ``reverse_complement`` returns a DNA sequence:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement(inplace=False)\n",
      " |      Seq('TCG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.reverse_complement_rna()\n",
      " |      MutableSeq('UCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.reverse_complement_rna(inplace=True)\n",
      " |      MutableSeq('UCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('UCG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``reverse_complement_rna`` is called on a ``Seq`` object with\n",
      " |      ``inplace=True``.\n",
      " |  \n",
      " |  rfind(self, sub, start=None, end=None)\n",
      " |      Return the highest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the highest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.rfind(\"AUG\")\n",
      " |      15\n",
      " |      \n",
      " |      The location of the typical start codon before that can be found by\n",
      " |      ending the search at position 15:\n",
      " |      \n",
      " |      >>> my_rna.rfind(\"AUG\", end=15)\n",
      " |      3\n",
      " |  \n",
      " |  rindex(self, sub, start=None, end=None)\n",
      " |      Return the highest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the highest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.rindex(\"AUG\")\n",
      " |      15\n",
      " |      \n",
      " |      The location of the typical start codon before that can be found by\n",
      " |      ending the search at position 15:\n",
      " |      \n",
      " |      >>> my_rna.rindex(\"AUG\", end=15)\n",
      " |      3\n",
      " |      \n",
      " |      This method performs the same search as the ``rfind`` method.  However,\n",
      " |      if the subsequence is not found, ``rfind`` returns -1 which ``rindex``\n",
      " |      raises a ValueError:\n",
      " |      \n",
      " |      >>> my_rna.rindex(\"T\")\n",
      " |      Traceback (most recent call last):\n",
      " |                 ...\n",
      " |      ValueError: ...\n",
      " |      >>> my_rna.rfind(\"T\")\n",
      " |      -1\n",
      " |  \n",
      " |  rsplit(self, sep=None, maxsplit=-1)\n",
      " |      Return a list of subsequences by splitting the sequence from the right.\n",
      " |      \n",
      " |      Return a list of the subsequences in the sequence (as Seq objects),\n",
      " |      using sep as the delimiter string.  If maxsplit is given, at\n",
      " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
      " |      splits are made.\n",
      " |      \n",
      " |      For consistency with the ``rsplit`` method of Python strings, any\n",
      " |      whitespace (tabs, spaces, newlines) is a separator if sep is None, the\n",
      " |      default value\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_aa = my_rna.translate()\n",
      " |      >>> my_aa\n",
      " |      Seq('VMAIVMGR*KGAR*L')\n",
      " |      >>> for pep in my_aa.rsplit(\"*\"):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR')\n",
      " |      Seq('L')\n",
      " |      >>> for pep in my_aa.rsplit(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR*KGAR')\n",
      " |      Seq('L')\n",
      " |      \n",
      " |      See also the split method, which splits the sequence starting from the\n",
      " |      beginning:\n",
      " |      \n",
      " |      >>> for pep in my_aa.split(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR*L')\n",
      " |  \n",
      " |  rstrip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, trailing whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.rstrip()\n",
      " |      Seq(' ACGT')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      from the trailing end instead.  The order of the characters to be\n",
      " |      removed is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGACGTTACG\").rstrip(\"GCA\")\n",
      " |      Seq('ACGACGTT')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.rstrip(inplace=False)\n",
      " |      MutableSeq(' ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.rstrip(inplace=True)\n",
      " |      MutableSeq(' ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``rstrip`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the strip and lstrip methods.\n",
      " |  \n",
      " |  split(self, sep=None, maxsplit=-1)\n",
      " |      Return a list of subsequences when splitting the sequence by separator sep.\n",
      " |      \n",
      " |      Return a list of the subsequences in the sequence (as Seq objects),\n",
      " |      using sep as the delimiter string.  If maxsplit is given, at\n",
      " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
      " |      splits are made.\n",
      " |      \n",
      " |      For consistency with the ``split`` method of Python strings, any\n",
      " |      whitespace (tabs, spaces, newlines) is a separator if sep is None, the\n",
      " |      default value\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_aa = my_rna.translate()\n",
      " |      >>> my_aa\n",
      " |      Seq('VMAIVMGR*KGAR*L')\n",
      " |      >>> for pep in my_aa.split(\"*\"):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR')\n",
      " |      Seq('L')\n",
      " |      >>> for pep in my_aa.split(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR*L')\n",
      " |      \n",
      " |      See also the rsplit method, which splits the sequence starting from the\n",
      " |      end:\n",
      " |      \n",
      " |      >>> for pep in my_aa.rsplit(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR*KGAR')\n",
      " |      Seq('L')\n",
      " |  \n",
      " |  startswith(self, prefix, start=None, end=None)\n",
      " |      Return True if the sequence starts with the given prefix, False otherwise.\n",
      " |      \n",
      " |      Return True if the sequence starts with the specified prefix\n",
      " |      (a string or another Seq object), False otherwise.\n",
      " |      With optional start, test sequence beginning at that position.\n",
      " |      With optional end, stop comparing sequence at that position.\n",
      " |      prefix can also be a tuple of strings to try.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.startswith(\"GUC\")\n",
      " |      True\n",
      " |      >>> my_rna.startswith(\"AUG\")\n",
      " |      False\n",
      " |      >>> my_rna.startswith(\"AUG\", 3)\n",
      " |      True\n",
      " |      >>> my_rna.startswith((\"UCC\", \"UCA\", \"UCG\"), 1)\n",
      " |      True\n",
      " |  \n",
      " |  strip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with leading and trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, leading and trailing whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.strip()\n",
      " |      Seq('ACGT')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      instead.  The order of the characters to be removed is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGTACGT\").strip(\"TGCA\")\n",
      " |      Seq('')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.strip(inplace=False)\n",
      " |      MutableSeq('ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.strip(inplace=True)\n",
      " |      MutableSeq('ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq('ACGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if ``strip``\n",
      " |      is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the lstrip and rstrip methods.\n",
      " |  \n",
      " |  transcribe(self, inplace=False)\n",
      " |      Transcribe a DNA sequence into RNA and return the RNA sequence as a new Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> coding_dna\n",
      " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> coding_dna.transcribe()\n",
      " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> sequence = MutableSeq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence.transcribe()\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      >>> sequence.transcribe(inplace=True)\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``transcribe`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      Trying to transcribe an RNA sequence has no effect.\n",
      " |      If you have a nucleotide sequence which might be DNA or RNA\n",
      " |      (or even a mixture), calling the transcribe method will ensure\n",
      " |      any T becomes U.\n",
      " |      \n",
      " |      Trying to transcribe a protein sequence will replace any\n",
      " |      T for Threonine with U for Selenocysteine, which has no\n",
      " |      biologically plausible rational.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_protein = Seq(\"MAIVMGRT\")\n",
      " |      >>> my_protein.transcribe()\n",
      " |      Seq('MAIVMGRU')\n",
      " |  \n",
      " |  translate(self, table='Standard', stop_symbol='*', to_stop=False, cds=False, gap='-')\n",
      " |      Turn a nucleotide sequence into a protein sequence by creating a new sequence object.\n",
      " |      \n",
      " |      This method will translate DNA or RNA sequences. It should not\n",
      " |      be used on protein sequences as any result will be biologically\n",
      " |      meaningless.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - table - Which codon table to use?  This can be either a name\n",
      " |         (string), an NCBI identifier (integer), or a CodonTable\n",
      " |         object (useful for non-standard genetic codes).  This\n",
      " |         defaults to the \"Standard\" table.\n",
      " |       - stop_symbol - Single character string, what to use for\n",
      " |         terminators.  This defaults to the asterisk, \"*\".\n",
      " |       - to_stop - Boolean, defaults to False meaning do a full\n",
      " |         translation continuing on past any stop codons (translated as the\n",
      " |         specified stop_symbol).  If True, translation is terminated at\n",
      " |         the first in frame stop codon (and the stop_symbol is not\n",
      " |         appended to the returned protein sequence).\n",
      " |       - cds - Boolean, indicates this is a complete CDS.  If True,\n",
      " |         this checks the sequence starts with a valid alternative start\n",
      " |         codon (which will be translated as methionine, M), that the\n",
      " |         sequence length is a multiple of three, and that there is a\n",
      " |         single in frame stop codon at the end (this will be excluded\n",
      " |         from the protein sequence, regardless of the to_stop option).\n",
      " |         If these tests fail, an exception is raised.\n",
      " |       - gap - Single character string to denote symbol used for gaps.\n",
      " |         Defaults to the minus sign.\n",
      " |      \n",
      " |      A ``Seq`` object is returned if ``translate`` is called on a ``Seq``\n",
      " |      object; a ``MutableSeq`` object is returned if ``translate`` is called\n",
      " |      pn a ``MutableSeq`` object.\n",
      " |      \n",
      " |      e.g. Using the standard table:\n",
      " |      \n",
      " |      >>> coding_dna = Seq(\"GTGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> coding_dna.translate()\n",
      " |      Seq('VAIVMGR*KGAR*')\n",
      " |      >>> coding_dna.translate(stop_symbol=\"@\")\n",
      " |      Seq('VAIVMGR@KGAR@')\n",
      " |      >>> coding_dna.translate(to_stop=True)\n",
      " |      Seq('VAIVMGR')\n",
      " |      \n",
      " |      Now using NCBI table 2, where TGA is not a stop codon:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=2)\n",
      " |      Seq('VAIVMGRWKGAR*')\n",
      " |      >>> coding_dna.translate(table=2, to_stop=True)\n",
      " |      Seq('VAIVMGRWKGAR')\n",
      " |      \n",
      " |      In fact, GTG is an alternative start codon under NCBI table 2, meaning\n",
      " |      this sequence could be a complete CDS:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=2, cds=True)\n",
      " |      Seq('MAIVMGRWKGAR')\n",
      " |      \n",
      " |      It isn't a valid CDS under NCBI table 1, due to both the start codon\n",
      " |      and also the in frame stop codons:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=1, cds=True)\n",
      " |      Traceback (most recent call last):\n",
      " |          ...\n",
      " |      Bio.Data.CodonTable.TranslationError: First codon 'GTG' is not a start codon\n",
      " |      \n",
      " |      If the sequence has no in-frame stop codon, then the to_stop argument\n",
      " |      has no effect:\n",
      " |      \n",
      " |      >>> coding_dna2 = Seq(\"TTGGCCATTGTAATGGGCCGC\")\n",
      " |      >>> coding_dna2.translate()\n",
      " |      Seq('LAIVMGR')\n",
      " |      >>> coding_dna2.translate(to_stop=True)\n",
      " |      Seq('LAIVMGR')\n",
      " |      \n",
      " |      NOTE - Ambiguous codons like \"TAN\" or \"NNN\" could be an amino acid\n",
      " |      or a stop codon.  These are translated as \"X\".  Any invalid codon\n",
      " |      (e.g. \"TA?\" or \"T-A\") will throw a TranslationError.\n",
      " |      \n",
      " |      NOTE - This does NOT behave like the python string's translate\n",
      " |      method.  For that use str(my_seq).translate(...) instead\n",
      " |  \n",
      " |  upper(self, inplace=False)\n",
      " |      Return the sequence in upper case.\n",
      " |      \n",
      " |      An upper-case copy of the sequence is returned if inplace is False,\n",
      " |      the default value:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_seq = Seq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      Seq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      Seq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      \n",
      " |      >>> my_seq.lower(inplace=True)\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper(inplace=True)\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``upper`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the ``lower`` method.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Readonly properties inherited from _SeqAbstractBaseClass:\n",
      " |  \n",
      " |  defined\n",
      " |      Return True if the sequence is defined, False if undefined or partially defined.\n",
      " |      \n",
      " |      Zero-length sequences are always considered to be defined.\n",
      " |  \n",
      " |  defined_ranges\n",
      " |      Return a tuple of the ranges where the sequence contents is defined.\n",
      " |      \n",
      " |      The return value has the format ((start1, end1), (start2, end2), ...).\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from _SeqAbstractBaseClass:\n",
      " |  \n",
      " |  __array_ufunc__ = None\n",
      "\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(dir(seq))\n",
    "print(help(seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5838b2e",
   "metadata": {},
   "source": [
    "Subscripting methods are available as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e4b0661b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A\n",
      "AC\n",
      "GTCACATGA\n"
     ]
    }
   ],
   "source": [
    "print(seq[3])\n",
    "print(seq[3:5])\n",
    "print(seq[::-1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae0a6c41",
   "metadata": {},
   "source": [
    "If needed, the `Seq` object can be converted into a string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "481041b7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGTACACTG\n",
      "XGTXCXCTG\n",
      "XGTXCXCTG\n"
     ]
    }
   ],
   "source": [
    "print(str(seq))\n",
    "print(str(seq).translate({65: 88}))\n",
    "print(str(seq).replace('A', 'X'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86986b39",
   "metadata": {},
   "source": [
    "To parse sequence from a file, you can use [Bio.SeqIO](https://biopython.org/docs/latest/api/Bio.SeqIO.html). [Here](https://biopython.org/wiki/SeqIO#file-formats) is the list of supported formats. The format name is passed into the `parse` method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fc17acba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sp|P0DTD1|R1AB_SARS2\n",
      "Seq('MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHLKDGTCGLV...VNN')\n",
      "7096\n",
      "sp|P0DTC1|R1A_SARS2\n",
      "Seq('MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHLKDGTCGLV...FAV')\n",
      "4405\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "sars2_it = SeqIO.parse(\"R1A-B_SARS2.fasta\", \"fasta\")\n",
    "for seq_record in sars2_it:    \n",
    "    print(seq_record.id)\n",
    "    print(repr(seq_record.seq))\n",
    "    print(len(seq_record))\n",
    "sars2_seq_recs = list(sars2_it)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46282c4c",
   "metadata": {},
   "source": [
    "The result is an iterator of [SeqRecord](https://biopython.org/docs/latest/api/Bio.SeqRecord.html)s. Other attributes of `SeqRecord` such as features or annotations are more relevant for other formats, such as genbank. The underlying gene for the two isoforms (R1A_SARS2/P0DTC1 and R1AB_SARS2/P0DTD1) is ORF1ab and the two isoforms are caused by ribosomal slippage during translation (see, e.g., [here](https://www.science.org/doi/full/10.1126/science.abf3546)). Both reading frames R1A_SARS2 and R1AB_SARS2 are polyproteins and are encoded by the same [gene](https://www.ncbi.nlm.nih.gov/gene/43740578). Let's explore this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "502a7c6f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_045512.2\n"
     ]
    }
   ],
   "source": [
    "gb_rec = list(SeqIO.parse(\"NC_045512.gb\", \"genbank\"))[0]\n",
    "print(gb_rec.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "934399c5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'molecule_type': 'ss-RNA', 'topology': 'linear', 'data_file_division': 'VRL', 'date': '18-JUL-2020', 'accessions': ['NC_045512'], 'sequence_version': 2, 'keywords': ['RefSeq'], 'source': 'Severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2)', 'organism': 'Severe acute respiratory syndrome coronavirus 2', 'taxonomy': ['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus'], 'references': [Reference(title='A new coronavirus associated with human respiratory disease in China', ...), Reference(title='Programmed ribosomal frameshifting in decoding the SARS-CoV genome', ...), Reference(title='The structure of a rigorously conserved RNA element within the SARS virus genome', ...), Reference(title=\"A phylogenetically conserved hairpin-type 3' untranslated region pseudoknot functions in coronavirus RNA replication\", ...), Reference(title='Direct Submission', ...), Reference(title='Direct Submission', ...)], 'comment': \"REVIEWED REFSEQ: This record has been curated by NCBI staff. The\\nreference sequence is identical to MN908947.\\nOn Jan 17, 2020 this sequence version replaced NC_045512.1.\\nAnnotation was added using homology to SARSr-CoV NC_004718.3. ###\\nFormerly called 'Wuhan seafood market pneumonia virus.' If you have\\nquestions or suggestions, please email us at info@ncbi.nlm.nih.gov\\nand include the accession number NC_045512.### Protein structures\\ncan be found at\\nhttps://www.ncbi.nlm.nih.gov/structure/?term=sars-cov-2.### Find\\nall other Severe acute respiratory syndrome coronavirus 2\\n(SARS-CoV-2) sequences at\\nhttps://www.ncbi.nlm.nih.gov/genbank/sars-cov-2-seqs/\\nCOMPLETENESS: full length.\", 'structured_comment': defaultdict(<class 'dict'>, {'Assembly-Data': {'Assembly Method': 'Megahit v. V1.1.3', 'Sequencing Technology': 'Illumina'}})}\n",
      "[SeqFeature(SimpleLocation(ExactPosition(0), ExactPosition(29903), strand=1), type='source', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(0), ExactPosition(265), strand=1), type=\"5'UTR\"), SeqFeature(SimpleLocation(ExactPosition(265), ExactPosition(21555), strand=1), type='gene', qualifiers=...), SeqFeature(CompoundLocation([SimpleLocation(ExactPosition(265), ExactPosition(13468), strand=1), SimpleLocation(ExactPosition(13467), ExactPosition(21555), strand=1)], 'join'), type='CDS', location_operator='join', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(265), ExactPosition(805), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(805), ExactPosition(2719), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2719), ExactPosition(8554), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(8554), ExactPosition(10054), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(10054), ExactPosition(10972), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(10972), ExactPosition(11842), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(11842), ExactPosition(12091), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(12091), ExactPosition(12685), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(12685), ExactPosition(13024), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(13024), ExactPosition(13441), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(CompoundLocation([SimpleLocation(ExactPosition(13441), ExactPosition(13468), strand=1), SimpleLocation(ExactPosition(13467), ExactPosition(16236), strand=1)], 'join'), type='mat_peptide', location_operator='join', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(16236), ExactPosition(18039), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(18039), ExactPosition(19620), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(19620), ExactPosition(20658), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(20658), ExactPosition(21552), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(265), ExactPosition(13483), strand=1), type='CDS', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(265), ExactPosition(805), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(805), ExactPosition(2719), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(2719), ExactPosition(8554), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(8554), ExactPosition(10054), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(10054), ExactPosition(10972), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(10972), ExactPosition(11842), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(11842), ExactPosition(12091), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(12091), ExactPosition(12685), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(12685), ExactPosition(13024), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(13024), ExactPosition(13441), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(13441), ExactPosition(13480), strand=1), type='mat_peptide', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(13475), ExactPosition(13503), strand=1), type='stem_loop', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(13487), ExactPosition(13542), strand=1), type='stem_loop', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(21562), ExactPosition(25384), strand=1), type='gene', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(21562), ExactPosition(25384), strand=1), type='CDS', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(25392), ExactPosition(26220), strand=1), type='gene', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(25392), ExactPosition(26220), strand=1), type='CDS', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(26244), ExactPosition(26472), strand=1), type='gene', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(26244), ExactPosition(26472), strand=1), type='CDS', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(26522), ExactPosition(27191), strand=1), type='gene', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(26522), ExactPosition(27191), strand=1), type='CDS', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(27201), ExactPosition(27387), strand=1), type='gene', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(27201), ExactPosition(27387), strand=1), type='CDS', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(27393), ExactPosition(27759), strand=1), type='gene', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(27393), ExactPosition(27759), strand=1), type='CDS', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(27755), ExactPosition(27887), strand=1), type='gene', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(27755), ExactPosition(27887), strand=1), type='CDS', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(27893), ExactPosition(28259), strand=1), type='gene', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(27893), ExactPosition(28259), strand=1), type='CDS', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(28273), ExactPosition(29533), strand=1), type='gene', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(28273), ExactPosition(29533), strand=1), type='CDS', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(29557), ExactPosition(29674), strand=1), type='gene', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(29557), ExactPosition(29674), strand=1), type='CDS', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(29608), ExactPosition(29644), strand=1), type='stem_loop', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(29628), ExactPosition(29657), strand=1), type='stem_loop', qualifiers=...), SeqFeature(SimpleLocation(ExactPosition(29674), ExactPosition(29903), strand=1), type=\"3'UTR\"), SeqFeature(SimpleLocation(ExactPosition(29727), ExactPosition(29768), strand=1), type='stem_loop', qualifiers=...)]\n"
     ]
    }
   ],
   "source": [
    "print(gb_rec.annotations)\n",
    "print(gb_rec.features)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13332305",
   "metadata": {},
   "source": [
    "Let's obtain all CDS (coding sequence) features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "d7512a84",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds = [seq_feature for seq_feature in gb_rec.features if seq_feature.type == 'CDS']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8acaa1b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqFeature(CompoundLocation([SimpleLocation(ExactPosition(265), ExactPosition(13468), strand=1), SimpleLocation(ExactPosition(13467), ExactPosition(21555), strand=1)], 'join'), type='CDS', location_operator='join', qualifiers=...),\n",
       " SeqFeature(SimpleLocation(ExactPosition(265), ExactPosition(13483), strand=1), type='CDS', qualifiers=...),\n",
       " SeqFeature(SimpleLocation(ExactPosition(21562), ExactPosition(25384), strand=1), type='CDS', qualifiers=...),\n",
       " SeqFeature(SimpleLocation(ExactPosition(25392), ExactPosition(26220), strand=1), type='CDS', qualifiers=...),\n",
       " SeqFeature(SimpleLocation(ExactPosition(26244), ExactPosition(26472), strand=1), type='CDS', qualifiers=...),\n",
       " SeqFeature(SimpleLocation(ExactPosition(26522), ExactPosition(27191), strand=1), type='CDS', qualifiers=...),\n",
       " SeqFeature(SimpleLocation(ExactPosition(27201), ExactPosition(27387), strand=1), type='CDS', qualifiers=...),\n",
       " SeqFeature(SimpleLocation(ExactPosition(27393), ExactPosition(27759), strand=1), type='CDS', qualifiers=...),\n",
       " SeqFeature(SimpleLocation(ExactPosition(27755), ExactPosition(27887), strand=1), type='CDS', qualifiers=...),\n",
       " SeqFeature(SimpleLocation(ExactPosition(27893), ExactPosition(28259), strand=1), type='CDS', qualifiers=...),\n",
       " SeqFeature(SimpleLocation(ExactPosition(28273), ExactPosition(29533), strand=1), type='CDS', qualifiers=...),\n",
       " SeqFeature(SimpleLocation(ExactPosition(29557), ExactPosition(29674), strand=1), type='CDS', qualifiers=...)]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "70edfa5c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['__bool__', '__class__', '__contains__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_flip', '_get_location_operator', '_get_ref', '_get_ref_db', '_get_strand', '_set_location_operator', '_set_ref', '_set_ref_db', '_set_strand', '_shift', 'extract', 'id', 'location', 'location_operator', 'qualifiers', 'ref', 'ref_db', 'strand', 'translate', 'type']\n"
     ]
    }
   ],
   "source": [
    "print(dir(cds[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "19127aa9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "   \"gene\": [\n",
      "      \"ORF1ab\"\n",
      "   ],\n",
      "   \"locus_tag\": [\n",
      "      \"GU280_gp01\"\n",
      "   ],\n",
      "   \"note\": [\n",
      "      \"pp1a\"\n",
      "   ],\n",
      "   \"codon_start\": [\n",
      "      \"1\"\n",
      "   ],\n",
      "   \"product\": [\n",
      "      \"ORF1a polyprotein\"\n",
      "   ],\n",
      "   \"protein_id\": [\n",
      "      \"YP_009725295.1\"\n",
      "   ],\n",
      "   \"db_xref\": [\n",
      "      \"GeneID:43740578\"\n",
      "   ],\n",
      "   \"translation\": [\n",
      "      \"MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHLKDGTCGLVEVEKGVLPQLEQPYVFIKRSDARTAPHGHVMVELVAELEGIQYGRSGETLGVLVPHVGEIPVAYRKVLLRKNGNKGAGGHSYGADLKSFDLGDELGTDPYEDFQENWNTKHSSGVTRELMRELNGGAYTRYVDNNFCGPDGYPLECIKDLLARAGKASCTLSEQLDFIDTKRGVYCCREHEHEIAWYTERSEKSYELQTPFEIKLAKKFDTFNGECPNFVFPLNSIIKTIQPRVEKKKLDGFMGRIRSVYPVASPNECNQMCLSTLMKCDHCGETSWQTGDFVKATCEFCGTENLTKEGATTCGYLPQNAVVKIYCPACHNSEVGPEHSLAEYHNESGLKTILRKGGRTIAFGGCVFSYVGCHNKCAYWVPRASANIGCNHTGVVGEGSEGLNDNLLEILQKEKVNINIVGDFKLNEEIAIILASFSASTSAFVETVKGLDYKAFKQIVESCGNFKVTKGKAKKGAWNIGEQKSILSPLYAFASEAARVVRSIFSRTLETAQNSVRVLQKAAITILDGISQYSLRLIDAMMFTSDLATNNLVVMAYITGGVVQLTSQWLTNIFGTVYEKLKPVLDWLEEKFKEGVEFLRDGWEIVKFISTCACEIVGGQIVTCAKEIKESVQTFFKLVNKFLALCADSIIIGGAKLKALNLGETFVTHSKGLYRKCVKSREETGLLMPLKAPKEIIFLEGETLPTEVLTEEVVLKTGDLQPLEQPTSEAVEAPLVGTPVCINGLMLLEIKDTEKYCALAPNMMVTNNTFTLKGGAPTKVTFGDDTVIEVQGYKSVNITFELDERIDKVLNEKCSAYTVELGTEVNEFACVVADAVIKTLQPVSELLTPLGIDLDEWSMATYYLFDESGEFKLASHMYCSFYPPDEDEEEGDCEEEEFEPSTQYEYGTEDDYQGKPLEFGATSAALQPEEEQEEDWLDDDSQQTVGQQDGSEDNQTTTIQTIVEVQPQLEMELTPVVQTIEVNSFSGYLKLTDNVYIKNADIVEEAKKVKPTVVVNAANVYLKHGGGVAGALNKATNNAMQVESDDYIATNGPLKVGGSCVLSGHNLAKHCLHVVGPNVNKGEDIQLLKSAYENFNQHEVLLAPLLSAGIFGADPIHSLRVCVDTVRTNVYLAVFDKNLYDKLVSSFLEMKSEKQVEQKIAEIPKEEVKPFITESKPSVEQRKQDDKKIKACVEEVTTTLEETKFLTENLLLYIDINGNLHPDSATLVSDIDITFLKKDAPYIVGDVVQEGVLTAVVIPTKKAGGTTEMLAKALRKVPTDNYITTYPGQGLNGYTVEEAKTVLKKCKSAFYILPSIISNEKQEILGTVSWNLREMLAHAEETRKLMPVCVETKAIVSTIQRKYKGIKIQEGVVDYGARFYFYTSKTTVASLINTLNDLNETLVTMPLGYVTHGLNLEEAARYMRSLKVPATVSVSSPDAVTAYNGYLTSSSKTPEEHFIETISLAGSYKDWSYSGQSTQLGIEFLKRGDKSVYYTSNPTTFHLDGEVITFDNLKTLLSLREVRTIKVFTTVDNINLHTQVVDMSMTYGQQFGPTYLDGADVTKIKPHNSHEGKTFYVLPNDDTLRVEAFEYYHTTDPSFLGRYMSALNHTKKWKYPQVNGLTSIKWADNNCYLATALLTLQQIELKFNPPALQDAYYRARAGEAANFCALILAYCNKTVGELGDVRETMSYLFQHANLDSCKRVLNVVCKTCGQQQTTLKGVEAVMYMGTLSYEQFKKGVQIPCTCGKQATKYLVQQESPFVMMSAPPAQYELKHGTFTCASEYTGNYQCGHYKHITSKETLYCIDGALLTKSSEYKGPITDVFYKENSYTTTIKPVTYKLDGVVCTEIDPKLDNYYKKDNSYFTEQPIDLVPNQPYPNASFDNFKFVCDNIKFADDLNQLTGYKKPASRELKVTFFPDLNGDVVAIDYKHYTPSFKKGAKLLHKPIVWHVNNATNKATYKPNTWCIRCLWSTKPVETSNSFDVLKSEDAQGMDNLACEDLKPVSEEVVENPTIQKDVLECNVKTTEVVGDIILKPANNSLKITEEVGHTDLMAAYVDNSSLTIKKPNELSRVLGLKTLATHGLAAVNSVPWDTIANYAKPFLNKVVSTTTNIVTRCLNRVCTNYMPYFFTLLLQLCTFTRSTNSRIKASMPTTIAKNTVKSVGKFCLEASFNYLKSPNFSKLINIIIWFLLLSVCLGSLIYSTAALGVLMSNLGMPSYCTGYREGYLNSTNVTIATYCTGSIPCSVCLSGLDSLDTYPSLETIQITISSFKWDLTAFGLVAEWFLAYILFTRFFYVLGLAAIMQLFFSYFAVHFISNSWLMWLIINLVQMAPISAMVRMYIFFASFYYVWKSYVHVVDGCNSSTCMMCYKRNRATRVECTTIVNGVRRSFYVYANGGKGFCKLHNWNCVNCDTFCAGSTFISDEVARDLSLQFKRPINPTDQSSYIVDSVTVKNGSIHLYFDKAGQKTYERHSLSHFVNLDNLRANNTKGSLPINVIVFDGKSKCEESSAKSASVYYSQLMCQPILLLDQALVSDVGDSAEVAVKMFDAYVNTFSSTFNVPMEKLKTLVATAEAELAKNVSLDNVLSTFISAARQGFVDSDVETKDVVECLKLSHQSDIEVTGDSCNNYMLTYNKVENMTPRDLGACIDCSARHINAQVAKSHNIALIWNVKDFMSLSEQLRKQIRSAAKKNNLPFKLTCATTRQVVNVVTTKIALKGGKIVNNWLKQLIKVTLVFLFVAAIFYLITPVHVMSKHTDFSSEIIGYKAIDGGVTRDIASTDTCFANKHADFDTWFSQRGGSYTNDKACPLIAAVITREVGFVVPGLPGTILRTTNGDFLHFLPRVFSAVGNICYTPSKLIEYTDFATSACVLAAECTIFKDASGKPVPYCYDTNVLEGSVAYESLRPDTRYVLMDGSIIQFPNTYLEGSVRVVTTFDSEYCRHGTCERSEAGVCVSTSGRWVLNNDYYRSLPGVFCGVDAVNLLTNMFTPLIQPIGALDISASIVAGGIVAIVVTCLAYYFMRFRRAFGEYSHVVAFNTLLFLMSFTVLCLTPVYSFLPGVYSVIYLYLTFYLTNDVSFLAHIQWMVMFTPLVPFWITIAYIICISTKHFYWFFSNYLKRRVVFNGVSFSTFEEAALCTFLLNKEMYLKLRSDVLLPLTQYNRYLALYNKYKYFSGAMDTTSYREAACCHLAKALNDFSNSGSDVLYQPPQTSITSAVLQSGFRKMAFPSGKVEGCMVQVTCGTTTLNGLWLDDVVYCPRHVICTSEDMLNPNYEDLLIRKSNHNFLVQAGNVQLRVIGHSMQNCVLKLKVDTANPKTPKYKFVRIQPGQTFSVLACYNGSPSGVYQCAMRPNFTIKGSFLNGSCGSVGFNIDYDCVSFCYMHHMELPTGVHAGTDLEGNFYGPFVDRQTAQAAGTDTTITVNVLAWLYAAVINGDRWFLNRFTTTLNDFNLVAMKYNYEPLTQDHVDILGPLSAQTGIAVLDMCASLKELLQNGMNGRTILGSALLEDEFTPFDVVRQCSGVTFQSAVKRTIKGTHHWLLLTILTSLLVLVQSTQWSLFFFLYENAFLPFAMGIIAMSAFAMMFVKHKHAFLCLFLLPSLATVAYFNMVYMPASWVMRIMTWLDMVDTSLSGFKLKDCVMYASAVVLLILMTARTVYDDGARRVWTLMNVLTLVYKVYYGNALDQAISMWALIISVTSNYSGVVTTVMFLARGIVFMCVEYCPIFFITGNTLQCIMLVYCFLGYFCTCYFGLFCLLNRYFRLTLGVYDYLVSTQEFRYMNSQGLLPPKNSIDAFKLNIKLLGVGGKPCIKVATVQSKMSDVKCTSVVLLSVLQQLRVESSSKLWAQCVQLHNDILLAKDTTEAFEKMVSLLSVLLSMQGAVDINKLCEEMLDNRATLQAIASEFSSLPSYAAFATAQEAYEQAVANGDSEVVLKKLKKSLNVAKSEFDRDAAMQRKLEKMADQAMTQMYKQARSEDKRAKVTSAMQTMLFTMLRKLDNDALNNIINNARDGCVPLNIIPLTTAAKLMVVIPDYNTYKNTCDGTTFTYASALWEIQQVVDADSKIVQLSEISMDNSPNLAWPLIVTALRANSAVKLQNNELSPVALRQMSCAAGTTQTACTDDNALAYYNTTKGGRFVLALLSDLQDLKWARFPKSDGTGTIYTELEPPCRFVTDTPKGPKVKYLYFIKGLNNLNRGMVLGSLAATVRLQAGNATEVPANSTVLSFCAFAVDAAKAYKDYLASGGQPITNCVKMLCTHTGTGQAITVTPEANMDQESFGGASCCLYCRCHIDHPNPKGFCDLKGKYVQIPTTCANDPVGFTLKNTVCTVCGMWKGYGCSCDQLREPMLQSADAQSFLNGFAV\"\n",
      "   ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "print(json.dumps(cds[1].qualifiers, indent=3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "43be2797",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHLKDGTCGLV...AV*')"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds[1].extract(gb_rec.seq).translate()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cc539a2",
   "metadata": {},
   "source": [
    "Now, let's get the DNA sequence for the the polyprotein 1ab."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "43a978f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MESLVPGFNE\n",
      "MESLVPGFNE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Python39\\lib\\site-packages\\Bio\\Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "aa_seq = cds[0].extract(gb_rec.seq).translate()\n",
    "print(aa_seq[:10])\n",
    "print(gb_rec.seq[265:].translate()[:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d17e0f0",
   "metadata": {},
   "source": [
    "To write a sequence into a file, use `SeqIO.write`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "b1008aec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SeqIO.write([gb_rec, SeqIO.SeqRecord(aa_seq, id=\"id\", description=\"aa\")], \"fasta_from_gb.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaac927e",
   "metadata": {},
   "source": [
    "### ---- Begin Exercise ----\n",
    "\n",
    "- Obtain the protein sequnece for polyprotein 1ab. Check with UniProt that it matches (just by eyeballing).\n",
    "- Obtain the protein sequence for the polyprotein 1a.\n",
    "- Obtain protein sequences for all the proteins and list them together with their names\n",
    "\n",
    "### ---- End Exercise ----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9f1ea87",
   "metadata": {},
   "source": [
    "#### Structures\n",
    "Structure processing is handled by the [Bio.PDB](https://biopython.org/docs/latest/api/Bio.PDB.html) package.\n",
    "\n",
    "To read a structure from a PDB file, use the `PDBParser`. We will be using the 3C-like proteinase protein, which is one of the processed proteins present in the ORF1a discussed above. One of it's structures is [7ALH](https://www.ebi.ac.uk/pdbe/entry/pdb/7alh). To see all the structures, I suggest checking out the PDBe-KB page for [P0DTD1](https://www.ebi.ac.uk/pdbe/pdbe-kb/proteins/P0DTD1)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "8b113320",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.PDB.PDBParser import PDBParser\n",
    "parser = PDBParser(PERMISSIVE=1)\n",
    "structure = parser.get_structure(\"7alh\", \"7alh.ent\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdcc527d",
   "metadata": {},
   "source": [
    "As the PDB format is considered deprecated, one should use the mmCIF file instead. This is done the same way as in case of PDB files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "035719e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.PDB.MMCIFParser import MMCIFParser\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"7alh\", \"7alh.cif\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81e46624",
   "metadata": {},
   "source": [
    "To retrieve the individual CIF dictionary fields, one can use the `MMCIF2Dict` module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ffb0783d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Crystal structure of the main protease (3CLpro/Mpro) of SARS-CoV-2 at 1.65A resolution (spacegroup C2).']\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB.MMCIFParser import MMCIF2Dict\n",
    "mmcif_dict = MMCIF2Dict(\"7alh.cif\")\n",
    "print(mmcif_dict[\"_citation.title\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95676e70",
   "metadata": {},
   "source": [
    "The structure record has the structure->model->chain->residue architecture.\n",
    "\n",
    "![SMRCA](http://biopython.org/DIST/docs/tutorial/images/smcra.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5852f822",
   "metadata": {},
   "source": [
    "Each of the levels in the hierarchy is represented by a submodule in Bio.PDB, namely [Bio.Structure](https://biopython.org/docs/latest/api/Bio.PDB.Structure.html), [Bio.Model](https://biopython.org/docs/latest/api/Bio.PDB.Model.html),[Bio.Chain](https://biopython.org/docs/latest/api/Bio.PDB.Chain.html),[Bio.Residue](https://biopython.org/docs/latest/api/Bio.PDB.Residue.html) and [Bio.Atom](https://biopython.org/docs/latest/api/Bio.PDB.Atom.html). For details regarding IDs, check the [section on ID](https://biopython.org/docs/1.75/api/Bio.PDB.Entity.html#Bio.PDB.Entity.Entity.get_full_id) of the Entity class which is the superclass of the Module/Chain/Residue/Atom classes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "bf3cac2c",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[<Model id=0>]\n",
      "---------- MODEL INFO ----------\n",
      "Full ID: ('7alh', 0)\n",
      "ID: 0\n",
      "[<Chain id=A>]\n",
      "---------- CHAIN INFO ----------\n",
      "Full ID: ('7alh', 0, 'A')\n",
      "ID: A\n",
      "[<Residue SER het=  resseq=1 icode= >, <Residue GLY het=  resseq=2 icode= >, <Residue PHE het=  resseq=3 icode= >, <Residue ARG het=  resseq=4 icode= >, <Residue LYS het=  resseq=5 icode= >, <Residue MET het=  resseq=6 icode= >, <Residue ALA het=  resseq=7 icode= >, <Residue PHE het=  resseq=8 icode= >, <Residue PRO het=  resseq=9 icode= >, <Residue SER het=  resseq=10 icode= >, <Residue GLY het=  resseq=11 icode= >, <Residue LYS het=  resseq=12 icode= >, <Residue VAL het=  resseq=13 icode= >, <Residue GLU het=  resseq=14 icode= >, <Residue GLY het=  resseq=15 icode= >, <Residue CYS het=  resseq=16 icode= >, <Residue MET het=  resseq=17 icode= >, <Residue VAL het=  resseq=18 icode= >, <Residue GLN het=  resseq=19 icode= >, <Residue VAL het=  resseq=20 icode= >, <Residue THR het=  resseq=21 icode= >, <Residue CYS het=  resseq=22 icode= >, <Residue GLY het=  resseq=23 icode= >, <Residue THR het=  resseq=24 icode= >, <Residue THR het=  resseq=25 icode= >, <Residue THR het=  resseq=26 icode= >, <Residue LEU het=  resseq=27 icode= >, <Residue ASN het=  resseq=28 icode= >, <Residue GLY het=  resseq=29 icode= >, <Residue LEU het=  resseq=30 icode= >, <Residue TRP het=  resseq=31 icode= >, <Residue LEU het=  resseq=32 icode= >, <Residue ASP het=  resseq=33 icode= >, <Residue ASP het=  resseq=34 icode= >, <Residue VAL het=  resseq=35 icode= >, <Residue VAL het=  resseq=36 icode= >, <Residue TYR het=  resseq=37 icode= >, <Residue CYS het=  resseq=38 icode= >, <Residue PRO het=  resseq=39 icode= >, <Residue ARG het=  resseq=40 icode= >, <Residue HIS het=  resseq=41 icode= >, <Residue VAL het=  resseq=42 icode= >, <Residue ILE het=  resseq=43 icode= >, <Residue CYS het=  resseq=44 icode= >, <Residue THR het=  resseq=45 icode= >, <Residue SER het=  resseq=46 icode= >, <Residue GLU het=  resseq=47 icode= >, <Residue ASP het=  resseq=48 icode= >, <Residue MET het=  resseq=49 icode= >, <Residue LEU het=  resseq=50 icode= >, <Residue ASN het=  resseq=51 icode= >, <Residue PRO het=  resseq=52 icode= >, <Residue ASN het=  resseq=53 icode= >, <Residue TYR het=  resseq=54 icode= >, <Residue GLU het=  resseq=55 icode= >, <Residue ASP het=  resseq=56 icode= >, <Residue LEU het=  resseq=57 icode= >, <Residue LEU het=  resseq=58 icode= >, <Residue ILE het=  resseq=59 icode= >, <Residue ARG het=  resseq=60 icode= >, <Residue LYS het=  resseq=61 icode= >, <Residue SER het=  resseq=62 icode= >, <Residue ASN het=  resseq=63 icode= >, <Residue HIS het=  resseq=64 icode= >, <Residue ASN het=  resseq=65 icode= >, <Residue PHE het=  resseq=66 icode= >, <Residue LEU het=  resseq=67 icode= >, <Residue VAL het=  resseq=68 icode= >, <Residue GLN het=  resseq=69 icode= >, <Residue ALA het=  resseq=70 icode= >, <Residue GLY het=  resseq=71 icode= >, <Residue ASN het=  resseq=72 icode= >, <Residue VAL het=  resseq=73 icode= >, <Residue GLN het=  resseq=74 icode= >, <Residue LEU het=  resseq=75 icode= >, <Residue ARG het=  resseq=76 icode= >, <Residue VAL het=  resseq=77 icode= >, <Residue ILE het=  resseq=78 icode= >, <Residue GLY het=  resseq=79 icode= >, <Residue HIS het=  resseq=80 icode= >, <Residue SER het=  resseq=81 icode= >, <Residue MET het=  resseq=82 icode= >, <Residue GLN het=  resseq=83 icode= >, <Residue ASN het=  resseq=84 icode= >, <Residue CYS het=  resseq=85 icode= >, <Residue VAL het=  resseq=86 icode= >, <Residue LEU het=  resseq=87 icode= >, <Residue LYS het=  resseq=88 icode= >, <Residue LEU het=  resseq=89 icode= >, <Residue LYS het=  resseq=90 icode= >, <Residue VAL het=  resseq=91 icode= >, <Residue ASP het=  resseq=92 icode= >, <Residue THR het=  resseq=93 icode= >, <Residue ALA het=  resseq=94 icode= >, <Residue ASN het=  resseq=95 icode= >, <Residue PRO het=  resseq=96 icode= >, <Residue LYS het=  resseq=97 icode= >, <Residue THR het=  resseq=98 icode= >, <Residue PRO het=  resseq=99 icode= >, <Residue LYS het=  resseq=100 icode= >, <Residue TYR het=  resseq=101 icode= >, <Residue LYS het=  resseq=102 icode= >, <Residue PHE het=  resseq=103 icode= >, <Residue VAL het=  resseq=104 icode= >, <Residue ARG het=  resseq=105 icode= >, <Residue ILE het=  resseq=106 icode= >, <Residue GLN het=  resseq=107 icode= >, <Residue PRO het=  resseq=108 icode= >, <Residue GLY het=  resseq=109 icode= >, <Residue GLN het=  resseq=110 icode= >, <Residue THR het=  resseq=111 icode= >, <Residue PHE het=  resseq=112 icode= >, <Residue SER het=  resseq=113 icode= >, <Residue VAL het=  resseq=114 icode= >, <Residue LEU het=  resseq=115 icode= >, <Residue ALA het=  resseq=116 icode= >, <Residue CYS het=  resseq=117 icode= >, <Residue TYR het=  resseq=118 icode= >, <Residue ASN het=  resseq=119 icode= >, <Residue GLY het=  resseq=120 icode= >, <Residue SER het=  resseq=121 icode= >, <Residue PRO het=  resseq=122 icode= >, <Residue SER het=  resseq=123 icode= >, <Residue GLY het=  resseq=124 icode= >, <Residue VAL het=  resseq=125 icode= >, <Residue TYR het=  resseq=126 icode= >, <Residue GLN het=  resseq=127 icode= >, <Residue CYS het=  resseq=128 icode= >, <Residue ALA het=  resseq=129 icode= >, <Residue MET het=  resseq=130 icode= >, <Residue ARG het=  resseq=131 icode= >, <Residue PRO het=  resseq=132 icode= >, <Residue ASN het=  resseq=133 icode= >, <Residue PHE het=  resseq=134 icode= >, <Residue THR het=  resseq=135 icode= >, <Residue ILE het=  resseq=136 icode= >, <Residue LYS het=  resseq=137 icode= >, <Residue GLY het=  resseq=138 icode= >, <Residue SER het=  resseq=139 icode= >, <Residue PHE het=  resseq=140 icode= >, <Residue LEU het=  resseq=141 icode= >, <Residue ASN het=  resseq=142 icode= >, <Residue GLY het=  resseq=143 icode= >, <Residue SER het=  resseq=144 icode= >, <Residue CYS het=  resseq=145 icode= >, <Residue GLY het=  resseq=146 icode= >, <Residue SER het=  resseq=147 icode= >, <Residue VAL het=  resseq=148 icode= >, <Residue GLY het=  resseq=149 icode= >, <Residue PHE het=  resseq=150 icode= >, <Residue ASN het=  resseq=151 icode= >, <Residue ILE het=  resseq=152 icode= >, <Residue ASP het=  resseq=153 icode= >, <Residue TYR het=  resseq=154 icode= >, <Residue ASP het=  resseq=155 icode= >, <Residue CYS het=  resseq=156 icode= >, <Residue VAL het=  resseq=157 icode= >, <Residue SER het=  resseq=158 icode= >, <Residue PHE het=  resseq=159 icode= >, <Residue CYS het=  resseq=160 icode= >, <Residue TYR het=  resseq=161 icode= >, <Residue MET het=  resseq=162 icode= >, <Residue HIS het=  resseq=163 icode= >, <Residue HIS het=  resseq=164 icode= >, <Residue MET het=  resseq=165 icode= >, <Residue GLU het=  resseq=166 icode= >, <Residue LEU het=  resseq=167 icode= >, <Residue PRO het=  resseq=168 icode= >, <Residue THR het=  resseq=169 icode= >, <Residue GLY het=  resseq=170 icode= >, <Residue VAL het=  resseq=171 icode= >, <Residue HIS het=  resseq=172 icode= >, <Residue ALA het=  resseq=173 icode= >, <Residue GLY het=  resseq=174 icode= >, <Residue THR het=  resseq=175 icode= >, <Residue ASP het=  resseq=176 icode= >, <Residue LEU het=  resseq=177 icode= >, <Residue GLU het=  resseq=178 icode= >, <Residue GLY het=  resseq=179 icode= >, <Residue ASN het=  resseq=180 icode= >, <Residue PHE het=  resseq=181 icode= >, <Residue TYR het=  resseq=182 icode= >, <Residue GLY het=  resseq=183 icode= >, <Residue PRO het=  resseq=184 icode= >, <Residue PHE het=  resseq=185 icode= >, <Residue VAL het=  resseq=186 icode= >, <Residue ASP het=  resseq=187 icode= >, <Residue ARG het=  resseq=188 icode= >, <Residue GLN het=  resseq=189 icode= >, <Residue THR het=  resseq=190 icode= >, <Residue ALA het=  resseq=191 icode= >, <Residue GLN het=  resseq=192 icode= >, <Residue ALA het=  resseq=193 icode= >, <Residue ALA het=  resseq=194 icode= >, <Residue GLY het=  resseq=195 icode= >, <Residue THR het=  resseq=196 icode= >, <Residue ASP het=  resseq=197 icode= >, <Residue THR het=  resseq=198 icode= >, <Residue THR het=  resseq=199 icode= >, <Residue ILE het=  resseq=200 icode= >, <Residue THR het=  resseq=201 icode= >, <Residue VAL het=  resseq=202 icode= >, <Residue ASN het=  resseq=203 icode= >, <Residue VAL het=  resseq=204 icode= >, <Residue LEU het=  resseq=205 icode= >, <Residue ALA het=  resseq=206 icode= >, <Residue TRP het=  resseq=207 icode= >, <Residue LEU het=  resseq=208 icode= >, <Residue TYR het=  resseq=209 icode= >, <Residue ALA het=  resseq=210 icode= >, <Residue ALA het=  resseq=211 icode= >, <Residue VAL het=  resseq=212 icode= >, <Residue ILE het=  resseq=213 icode= >, <Residue ASN het=  resseq=214 icode= >, <Residue GLY het=  resseq=215 icode= >, <Residue ASP het=  resseq=216 icode= >, <Residue ARG het=  resseq=217 icode= >, <Residue TRP het=  resseq=218 icode= >, <Residue PHE het=  resseq=219 icode= >, <Residue LEU het=  resseq=220 icode= >, <Residue ASN het=  resseq=221 icode= >, <Residue ARG het=  resseq=222 icode= >, <Residue PHE het=  resseq=223 icode= >, <Residue THR het=  resseq=224 icode= >, <Residue THR het=  resseq=225 icode= >, <Residue THR het=  resseq=226 icode= >, <Residue LEU het=  resseq=227 icode= >, <Residue ASN het=  resseq=228 icode= >, <Residue ASP het=  resseq=229 icode= >, <Residue PHE het=  resseq=230 icode= >, <Residue ASN het=  resseq=231 icode= >, <Residue LEU het=  resseq=232 icode= >, <Residue VAL het=  resseq=233 icode= >, <Residue ALA het=  resseq=234 icode= >, <Residue MET het=  resseq=235 icode= >, <Residue LYS het=  resseq=236 icode= >, <Residue TYR het=  resseq=237 icode= >, <Residue ASN het=  resseq=238 icode= >, <Residue TYR het=  resseq=239 icode= >, <Residue GLU het=  resseq=240 icode= >, <Residue PRO het=  resseq=241 icode= >, <Residue LEU het=  resseq=242 icode= >, <Residue THR het=  resseq=243 icode= >, <Residue GLN het=  resseq=244 icode= >, <Residue ASP het=  resseq=245 icode= >, <Residue HIS het=  resseq=246 icode= >, <Residue VAL het=  resseq=247 icode= >, <Residue ASP het=  resseq=248 icode= >, <Residue ILE het=  resseq=249 icode= >, <Residue LEU het=  resseq=250 icode= >, <Residue GLY het=  resseq=251 icode= >, <Residue PRO het=  resseq=252 icode= >, <Residue LEU het=  resseq=253 icode= >, <Residue SER het=  resseq=254 icode= >, <Residue ALA het=  resseq=255 icode= >, <Residue GLN het=  resseq=256 icode= >, <Residue THR het=  resseq=257 icode= >, <Residue GLY het=  resseq=258 icode= >, <Residue ILE het=  resseq=259 icode= >, <Residue ALA het=  resseq=260 icode= >, <Residue VAL het=  resseq=261 icode= >, <Residue LEU het=  resseq=262 icode= >, <Residue ASP het=  resseq=263 icode= >, <Residue MET het=  resseq=264 icode= >, <Residue CYS het=  resseq=265 icode= >, <Residue ALA het=  resseq=266 icode= >, <Residue SER het=  resseq=267 icode= >, <Residue LEU het=  resseq=268 icode= >, <Residue LYS het=  resseq=269 icode= >, <Residue GLU het=  resseq=270 icode= >, <Residue LEU het=  resseq=271 icode= >, <Residue LEU het=  resseq=272 icode= >, <Residue GLN het=  resseq=273 icode= >, <Residue ASN het=  resseq=274 icode= >, <Residue GLY het=  resseq=275 icode= >, <Residue MET het=  resseq=276 icode= >, <Residue ASN het=  resseq=277 icode= >, <Residue GLY het=  resseq=278 icode= >, <Residue ARG het=  resseq=279 icode= >, <Residue THR het=  resseq=280 icode= >, <Residue ILE het=  resseq=281 icode= >, <Residue LEU het=  resseq=282 icode= >, <Residue GLY het=  resseq=283 icode= >, <Residue SER het=  resseq=284 icode= >, <Residue ALA het=  resseq=285 icode= >, <Residue LEU het=  resseq=286 icode= >, <Residue LEU het=  resseq=287 icode= >, <Residue GLU het=  resseq=288 icode= >, <Residue ASP het=  resseq=289 icode= >, <Residue GLU het=  resseq=290 icode= >, <Residue PHE het=  resseq=291 icode= >, <Residue THR het=  resseq=292 icode= >, <Residue PRO het=  resseq=293 icode= >, <Residue PHE het=  resseq=294 icode= >, <Residue ASP het=  resseq=295 icode= >, <Residue VAL het=  resseq=296 icode= >, <Residue VAL het=  resseq=297 icode= >, <Residue ARG het=  resseq=298 icode= >, <Residue GLN het=  resseq=299 icode= >, <Residue CYS het=  resseq=300 icode= >, <Residue SER het=  resseq=301 icode= >, <Residue GLY het=  resseq=302 icode= >, <Residue VAL het=  resseq=303 icode= >, <Residue THR het=  resseq=304 icode= >, <Residue PHE het=  resseq=305 icode= >, <Residue HOH het=W resseq=401 icode= >, <Residue HOH het=W resseq=402 icode= >, <Residue HOH het=W resseq=403 icode= >, <Residue HOH het=W resseq=404 icode= >, <Residue HOH het=W resseq=405 icode= >, <Residue HOH het=W resseq=406 icode= >, <Residue HOH het=W resseq=407 icode= >, <Residue HOH het=W resseq=408 icode= >, <Residue HOH het=W resseq=409 icode= >, <Residue HOH het=W resseq=410 icode= >, <Residue HOH het=W resseq=411 icode= >, <Residue HOH het=W resseq=412 icode= >, <Residue HOH het=W resseq=413 icode= >, <Residue HOH het=W resseq=414 icode= >, <Residue HOH het=W resseq=415 icode= >, <Residue HOH het=W resseq=416 icode= >, <Residue HOH het=W resseq=417 icode= >, <Residue HOH het=W resseq=418 icode= >, <Residue HOH het=W resseq=419 icode= >, <Residue HOH het=W resseq=420 icode= >, <Residue HOH het=W resseq=421 icode= >, <Residue HOH het=W resseq=422 icode= >, <Residue HOH het=W resseq=423 icode= >, <Residue HOH het=W resseq=424 icode= >, <Residue HOH het=W resseq=425 icode= >, <Residue HOH het=W resseq=426 icode= >, <Residue HOH het=W resseq=427 icode= >, <Residue HOH het=W resseq=428 icode= >, <Residue HOH het=W resseq=429 icode= >, <Residue HOH het=W resseq=430 icode= >, <Residue HOH het=W resseq=431 icode= >, <Residue HOH het=W resseq=432 icode= >, <Residue HOH het=W resseq=433 icode= >, <Residue HOH het=W resseq=434 icode= >, <Residue HOH het=W resseq=435 icode= >, <Residue HOH het=W resseq=436 icode= >, <Residue HOH het=W resseq=437 icode= >, <Residue HOH het=W resseq=438 icode= >, <Residue HOH het=W resseq=439 icode= >, <Residue HOH het=W resseq=440 icode= >, <Residue HOH het=W resseq=441 icode= >, <Residue HOH het=W resseq=442 icode= >, <Residue HOH het=W resseq=443 icode= >, <Residue HOH het=W resseq=444 icode= >, <Residue HOH het=W resseq=445 icode= >, <Residue HOH het=W resseq=446 icode= >, <Residue HOH het=W resseq=447 icode= >, <Residue HOH het=W resseq=448 icode= >, <Residue HOH het=W resseq=449 icode= >, <Residue HOH het=W resseq=450 icode= >, <Residue HOH het=W resseq=451 icode= >, <Residue HOH het=W resseq=452 icode= >, <Residue HOH het=W resseq=453 icode= >, <Residue HOH het=W resseq=454 icode= >, <Residue HOH het=W resseq=455 icode= >, <Residue HOH het=W resseq=456 icode= >, <Residue HOH het=W resseq=457 icode= >, <Residue HOH het=W resseq=458 icode= >, <Residue HOH het=W resseq=459 icode= >, <Residue HOH het=W resseq=460 icode= >, <Residue HOH het=W resseq=461 icode= >, <Residue HOH het=W resseq=462 icode= >, <Residue HOH het=W resseq=463 icode= >, <Residue HOH het=W resseq=464 icode= >, <Residue HOH het=W resseq=465 icode= >, <Residue HOH het=W resseq=466 icode= >, <Residue HOH het=W resseq=467 icode= >, <Residue HOH het=W resseq=468 icode= >, <Residue HOH het=W resseq=469 icode= >, <Residue HOH het=W resseq=470 icode= >, <Residue HOH het=W resseq=471 icode= >, <Residue HOH het=W resseq=472 icode= >, <Residue HOH het=W resseq=473 icode= >, <Residue HOH het=W resseq=474 icode= >, <Residue HOH het=W resseq=475 icode= >, <Residue HOH het=W resseq=476 icode= >, <Residue HOH het=W resseq=477 icode= >, <Residue HOH het=W resseq=478 icode= >, <Residue HOH het=W resseq=479 icode= >, <Residue HOH het=W resseq=480 icode= >, <Residue HOH het=W resseq=481 icode= >, <Residue HOH het=W resseq=482 icode= >, <Residue HOH het=W resseq=483 icode= >, <Residue HOH het=W resseq=484 icode= >, <Residue HOH het=W resseq=485 icode= >, <Residue HOH het=W resseq=486 icode= >, <Residue HOH het=W resseq=487 icode= >, <Residue HOH het=W resseq=488 icode= >, <Residue HOH het=W resseq=489 icode= >, <Residue HOH het=W resseq=490 icode= >, <Residue HOH het=W resseq=491 icode= >, <Residue HOH het=W resseq=492 icode= >, <Residue HOH het=W resseq=493 icode= >, <Residue HOH het=W resseq=494 icode= >, <Residue HOH het=W resseq=495 icode= >, <Residue HOH het=W resseq=496 icode= >, <Residue HOH het=W resseq=497 icode= >, <Residue HOH het=W resseq=498 icode= >, <Residue HOH het=W resseq=499 icode= >, <Residue HOH het=W resseq=500 icode= >, <Residue HOH het=W resseq=501 icode= >, <Residue HOH het=W resseq=502 icode= >, <Residue HOH het=W resseq=503 icode= >, <Residue HOH het=W resseq=504 icode= >, <Residue HOH het=W resseq=505 icode= >, <Residue HOH het=W resseq=506 icode= >, <Residue HOH het=W resseq=507 icode= >, <Residue HOH het=W resseq=508 icode= >, <Residue HOH het=W resseq=509 icode= >, <Residue HOH het=W resseq=510 icode= >, <Residue HOH het=W resseq=511 icode= >, <Residue HOH het=W resseq=512 icode= >, <Residue HOH het=W resseq=513 icode= >, <Residue HOH het=W resseq=514 icode= >, <Residue HOH het=W resseq=515 icode= >, <Residue HOH het=W resseq=516 icode= >, <Residue HOH het=W resseq=517 icode= >, <Residue HOH het=W resseq=518 icode= >, <Residue HOH het=W resseq=519 icode= >, <Residue HOH het=W resseq=520 icode= >, <Residue HOH het=W resseq=521 icode= >, <Residue HOH het=W resseq=522 icode= >, <Residue HOH het=W resseq=523 icode= >, <Residue HOH het=W resseq=524 icode= >, <Residue HOH het=W resseq=525 icode= >, <Residue HOH het=W resseq=526 icode= >, <Residue HOH het=W resseq=527 icode= >, <Residue HOH het=W resseq=528 icode= >, <Residue HOH het=W resseq=529 icode= >, <Residue HOH het=W resseq=530 icode= >, <Residue HOH het=W resseq=531 icode= >, <Residue HOH het=W resseq=532 icode= >, <Residue HOH het=W resseq=533 icode= >, <Residue HOH het=W resseq=534 icode= >, <Residue HOH het=W resseq=535 icode= >, <Residue HOH het=W resseq=536 icode= >, <Residue HOH het=W resseq=537 icode= >, <Residue HOH het=W resseq=538 icode= >, <Residue HOH het=W resseq=539 icode= >, <Residue HOH het=W resseq=540 icode= >, <Residue HOH het=W resseq=541 icode= >, <Residue HOH het=W resseq=542 icode= >, <Residue HOH het=W resseq=543 icode= >, <Residue HOH het=W resseq=544 icode= >, <Residue HOH het=W resseq=545 icode= >, <Residue HOH het=W resseq=546 icode= >, <Residue HOH het=W resseq=547 icode= >, <Residue HOH het=W resseq=548 icode= >, <Residue HOH het=W resseq=549 icode= >, <Residue HOH het=W resseq=550 icode= >, <Residue HOH het=W resseq=551 icode= >, <Residue HOH het=W resseq=552 icode= >, <Residue HOH het=W resseq=553 icode= >, <Residue HOH het=W resseq=554 icode= >, <Residue HOH het=W resseq=555 icode= >, <Residue HOH het=W resseq=556 icode= >, <Residue HOH het=W resseq=557 icode= >, <Residue HOH het=W resseq=558 icode= >, <Residue HOH het=W resseq=559 icode= >, <Residue HOH het=W resseq=560 icode= >, <Residue HOH het=W resseq=561 icode= >, <Residue HOH het=W resseq=562 icode= >, <Residue HOH het=W resseq=563 icode= >, <Residue HOH het=W resseq=564 icode= >, <Residue HOH het=W resseq=565 icode= >, <Residue HOH het=W resseq=566 icode= >, <Residue HOH het=W resseq=567 icode= >, <Residue HOH het=W resseq=568 icode= >, <Residue HOH het=W resseq=569 icode= >, <Residue HOH het=W resseq=570 icode= >, <Residue HOH het=W resseq=571 icode= >, <Residue HOH het=W resseq=572 icode= >, <Residue HOH het=W resseq=573 icode= >, <Residue HOH het=W resseq=574 icode= >, <Residue HOH het=W resseq=575 icode= >, <Residue HOH het=W resseq=576 icode= >, <Residue HOH het=W resseq=577 icode= >, <Residue HOH het=W resseq=578 icode= >, <Residue HOH het=W resseq=579 icode= >, <Residue HOH het=W resseq=580 icode= >, <Residue HOH het=W resseq=581 icode= >, <Residue HOH het=W resseq=582 icode= >, <Residue HOH het=W resseq=583 icode= >, <Residue HOH het=W resseq=584 icode= >, <Residue HOH het=W resseq=585 icode= >, <Residue HOH het=W resseq=586 icode= >, <Residue HOH het=W resseq=587 icode= >, <Residue HOH het=W resseq=588 icode= >, <Residue HOH het=W resseq=589 icode= >, <Residue HOH het=W resseq=590 icode= >, <Residue HOH het=W resseq=591 icode= >, <Residue HOH het=W resseq=592 icode= >, <Residue HOH het=W resseq=593 icode= >, <Residue HOH het=W resseq=594 icode= >, <Residue HOH het=W resseq=595 icode= >, <Residue HOH het=W resseq=596 icode= >, <Residue HOH het=W resseq=597 icode= >, <Residue HOH het=W resseq=598 icode= >, <Residue HOH het=W resseq=599 icode= >, <Residue HOH het=W resseq=600 icode= >, <Residue HOH het=W resseq=601 icode= >, <Residue HOH het=W resseq=602 icode= >, <Residue HOH het=W resseq=603 icode= >, <Residue HOH het=W resseq=604 icode= >, <Residue HOH het=W resseq=605 icode= >, <Residue HOH het=W resseq=606 icode= >, <Residue HOH het=W resseq=607 icode= >, <Residue HOH het=W resseq=608 icode= >, <Residue HOH het=W resseq=609 icode= >, <Residue HOH het=W resseq=610 icode= >, <Residue HOH het=W resseq=611 icode= >, <Residue HOH het=W resseq=612 icode= >, <Residue HOH het=W resseq=613 icode= >, <Residue HOH het=W resseq=614 icode= >, <Residue HOH het=W resseq=615 icode= >, <Residue HOH het=W resseq=616 icode= >, <Residue HOH het=W resseq=617 icode= >, <Residue HOH het=W resseq=618 icode= >, <Residue HOH het=W resseq=619 icode= >, <Residue HOH het=W resseq=620 icode= >, <Residue HOH het=W resseq=621 icode= >, <Residue HOH het=W resseq=622 icode= >, <Residue HOH het=W resseq=623 icode= >, <Residue HOH het=W resseq=624 icode= >, <Residue HOH het=W resseq=625 icode= >, <Residue HOH het=W resseq=626 icode= >, <Residue HOH het=W resseq=627 icode= >, <Residue HOH het=W resseq=628 icode= >, <Residue HOH het=W resseq=629 icode= >, <Residue HOH het=W resseq=630 icode= >, <Residue HOH het=W resseq=631 icode= >, <Residue HOH het=W resseq=632 icode= >, <Residue HOH het=W resseq=633 icode= >, <Residue HOH het=W resseq=634 icode= >, <Residue HOH het=W resseq=635 icode= >, <Residue HOH het=W resseq=636 icode= >, <Residue HOH het=W resseq=637 icode= >, <Residue HOH het=W resseq=638 icode= >, <Residue HOH het=W resseq=639 icode= >, <Residue HOH het=W resseq=640 icode= >]\n"
     ]
    }
   ],
   "source": [
    "print(structure.get_list())\n",
    "\n",
    "print('---------- MODEL INFO ----------')\n",
    "\n",
    "model = structure[0]\n",
    "print(f\"Full ID: {model.get_full_id()}\\nID: {model.get_id()}\")\n",
    "print(model.get_list())\n",
    "\n",
    "print('---------- CHAIN INFO ----------')\n",
    "chain = model['A']\n",
    "print(f\"Full ID: {chain.get_full_id()}\\nID: {chain.get_id()}\")\n",
    "print(chain.get_list())\n",
    "\n",
    "# print('---------- RESIDUE INFO ----------')\n",
    "# res = chain[(' ',1,' ')]\n",
    "# print(f\"Full ID: {res.get_full_id()}\\nID: {res.get_id()}\")\n",
    "# print(res.get_resname())\n",
    "# res = chain[1]\n",
    "# print(res.get_resname())\n",
    "\n",
    "# print(res.get_list())\n",
    "# print('---------- ATOM INFO ----------')\n",
    "# atom=res['CA']\n",
    "# print(f\"Full ID: {atom.get_full_id()}\\nID: {atom.get_id()}\")\n",
    "# print(f\"{atom.get_name()}\\n{atom.get_id()}\\n{atom.get_coord()}\\n{atom.get_fullname()}\")\n",
    "# print(atom.get_vector())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "414a65bb",
   "metadata": {},
   "source": [
    "In the above script, notice that the residue ID is a triplet where the first position stores the residue type ('H' for hetero atoms, 'W' for water and ' ' for everything else), the second its number and the last position is the insertion code. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aac2e98b",
   "metadata": {},
   "source": [
    "To download a file from PDB, one can use the PDBList module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "bc8ca76f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Structure exists: '.\\7lkr.cif' \n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB.PDBList import PDBList\n",
    "pdbl = PDBList()\n",
    "pbl_7lkr=pdbl.retrieve_pdb_file(\"7LKR\", file_format=\"mmCif\", pdir=\".\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e674be0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Python39\\lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 4467.\n",
      "  warnings.warn(\n",
      "c:\\Python39\\lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 4531.\n",
      "  warnings.warn(\n",
      "c:\\Python39\\lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 4608.\n",
      "  warnings.warn(\n",
      "c:\\Python39\\lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 4774.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB.MMCIFParser import MMCIFParser\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"7lkr\", \"7lkr.cif\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d77eb087",
   "metadata": {},
   "source": [
    "### ---- Begin Exercise ----\n",
    "\n",
    "- Iterate over all atoms of the structure\n",
    "- List all water residues (the first field of the residue id is 'W')\n",
    "- How many water molecules are in the recrod?\n",
    "- How many heteroatoms are there in the recod (the first field of the residue id is 'H').\n",
    "- Find a structure in PDB with at least one ligand (different from water) and write a code which lists all the ligands. (All such ligand can be found in `HETNAM` sections in PDB and in `_chem_comp.id` records in mmCIF).\n",
    "\n",
    "### ---- End Exercise ----"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.5 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  },
  "vscode": {
   "interpreter": {
    "hash": "81794d4967e6c3204c66dcd87b604927b115b27c00565d3d43f05ba2f3a2cb0d"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
